feat: add tri-state UI settings pages and auto opt-in preference

eunjae-lee opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- fixes CAL-6930

This PR implements the feature opt-in settings UI across user, team, and organization levels, enabling administrators and users to control experimental feature adoption with granular permissions.

## Summary

- Adds tri-state toggle UI (On/Off/Use default for users; Allow/Block/Let users decide for teams/orgs) for managing feature opt-in preferences
- Implements auto opt-in preference allowing automatic enrollment in future experimental features
- Creates reusable PBAC (Permission-Based Access Control) procedure utilities to reduce boilerplate
- **Adds per-feature policy support (`permissive`/`strict`) for controlling opt-in behavior**

## Before testing

1. Uncomment packages/features/feature-opt-in/config.ts:15-19
2. Run this query: `update "Feature" set enabled = true where slug='bookings-v3';` (it's false by default, meaning globally explicitly disabled)

## Demo

part 1 / 2


https://github.com/user-attachments/assets/a4ed46f2-86ce-49ef-933d-e8f667bf3075



part 2 / 2


https://github.com/user-attachments/assets/a603114c-3097-4ded-aa0e-0ac71dc64d52



## Changes

### New Settings Pages
| Level | Route | Description |
|-------|-------|-------------|
| User | `/settings/my-account/features` | Personal feature opt-in settings |
| Team | `/settings/teams/[id]/features` | Team-level feature controls |
| Organization | `/settings/organizations/features` | Org-wide feature policies |

### Architecture

**UI Components**
- `FeaturesSettings.tsx` - Shared component handling tri-state ToggleGroup for all scopes
- Displays blocked warnings when user enables a feature that's disabled by team/org
- **Disables toggle controls** when feature is blocked by higher level (org/team)
- **Accepts `canEdit` prop** to disable all toggles when user lacks update permission

**React Hooks**
- `useUserFeatureOptIn` - Manages user-level feature states with org/team blocking detection
- `useTeamFeatureOptIn` - Manages team-level feature states with org blocking detection
- `useOrganizationFeatureOptIn` - Manages organization-level feature states

**tRPC Endpoints (6 new)**
- `getUserAutoOptIn` / `setUserAutoOptIn`
- `getTeamAutoOptIn` / `setTeamAutoOptIn`
- `getOrganizationAutoOptIn` / `setOrganizationAutoOptIn`

**PBAC Utilities**
- `createTeamPbacProcedure` - Reusable procedure for team-scoped permission checks
- `createOrgPbacProcedure` - Reusable procedure for org-scoped permission checks with typed context extension

### Navigation
- Features tab conditionally shown in settings sidebar only when `HAS_OPT_IN_FEATURES` is true

## Updates since last revision

- Added "disabled by organization" badge for Team Level Features page
- Added unit tests for PBAC utility procedures (`pbacProcedures.test.ts`)
- Added unit tests for `FeatureOptInService.listFeaturesForTeam` (`FeatureOptInService.test.ts`)
- Refactored tRPC router to use `TeamRepository.findParentOrganizationByTeamId` instead of direct prisma calls
- Refactored `createMutationCallbacks` to `useMutationCallbacks` hook pattern
- Added `isBlockedByHigherLevel` function to disable toggles when blocked
- Added per-feature policy support (`permissive`/`strict`)
  - Added `OptInFeaturePolicy` type with two modes to `OptInFeatureConfig`
  - `permissive` (default): user opt-in can activate; disables only win if ALL teams disable
  - `strict`: user opt-in alone not enough; ANY explicit disable blocks
  - Added new `EffectiveStateReason` values: `feature_any_team_disabled`, `feature_user_only_not_allowed`
  - Updated `computeEffectiveStateAcrossTeams` to handle both policies
- Rewrote test file with comprehensive scenario documentation tables
  - Added ASCII tables documenting all 10 permissive and 10 strict policy scenarios
  - Added key differences table comparing policy behaviors
  - Made `policy` parameter required (no longer optional with default)
- Moved PBAC procedures to `packages/trpc/server/procedures/pbacProcedures.ts` for better reusability
  - Renamed `util.ts` → `pbacProcedures.ts` and moved to general procedures folder
  - Renamed `util.test.ts` → `pbacProcedures.test.ts` alongside the main file
  - Updated imports in `_router.ts` to use new location
- Improved middleware typing to avoid non-null assertions
  - Removed explicit return type from `createOrgPbacProcedure` to let TypeScript infer the extended context type
  - Handlers now use `ctx.organizationId` directly (typed as `number`) instead of `ctx.user.organizationId!`
  - The middleware validates `organizationId` and adds it to the context, so downstream handlers can safely access it
- Updated to use `featureOptIn.read` and `featureOptIn.update` permissions (from PR #26761) instead of generic `team.*` and `organization.*` permissions
  - Router endpoints now use `featureOptIn.read` for read operations and `featureOptIn.update` for write operations
  - Organization features page now uses `validateUserHasOrg` + `getResourcePermissions` with `Resource.FeatureOptIn` instead of `validateUserHasOrgAdmin`
- **Latest:** Added PBAC to team features page and `canEdit` prop to views
  - Team features page now uses `getResourcePermissions` with `Resource.FeatureOptIn` for permission checks
  - Both org and team features pages now pass `canEdit` to their view components
  - `FeaturesSettings` component accepts `canEdit` prop (defaults to `true`) to disable all toggles when user lacks update permission
  - Auto opt-in toggle is also disabled when `canEdit` is false


## Testing Instructions

1. Navigate to each settings page:
   - `/settings/my-account/features`
   - `/settings/teams/[id]/features`
   - `/settings/organizations/features`
2. Test tri-state toggles (On/Off/Use default or Allow/Block/Let users decide)
3. Test auto opt-in checkbox at each level
4. Verify warning icon appears when user enables a feature blocked by team/org
5. Verify warning icon appears on team features page when team enables a feature blocked by org
6. **Verify toggle is disabled (not just showing warning) when feature is blocked by higher level**
7. **Test with a user who has `featureOptIn.read` but not `featureOptIn.update` permission - verify they can view but not edit**
8. Run unit tests:
   - `yarn test packages/trpc/server/procedures/pbacProcedures.test.ts`
   - `TZ=UTC yarn test packages/features/feature-opt-in/services/FeatureOptInService.test.ts`
   - `TZ=UTC yarn test packages/features/feature-opt-in/lib/computeEffectiveState.test.ts`

## Visual Demo (For contributors especially)



## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. Run the migration: `yarn prisma migrate dev`
3. Navigate to user settings and verify the features settings page renders
4. Navigate to team settings and verify the features settings page renders
5. Navigate to organization settings and verify the features settings page renders
6. Test the tri-state toggle for each level (user/team/org)
7. Test the auto opt-in checkbox for each level
8. Verify warning appears when user enables a feature that team/org has blocked
9. Verify warning appears on team features page when team enables a feature that org has blocked
10. **Verify toggle controls are disabled when org/team has blocked the feature**
11. **Verify toggles are disabled when user has read but not update permission**

**Note:** The settings view components are created but may need route registration to be accessible in the app.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings

## Human Review Checklist

- [ ] **Dependency check**: Ensure PR #26761 is merged (adds FeatureOptIn resource to permission registry with `featureOptIn.read` and `featureOptIn.update` permissions)
- [ ] Verify routes are properly set up to access these settings pages (may need additional route registration)
- [ ] Review the `useMutationCallbacks` hook pattern - verify memoization with `useMemo` and `useCallback` is correct
- [ ] **Verify `createOrgPbacProcedure` middleware context extension works correctly** - handlers use `ctx.organizationId` (typed as `number` via TypeScript inference from middleware)
- [ ] Review the conditional TRPC hook usage pattern in `TeamOrgFeaturesSettings.tsx` - hooks are selected based on `entityType` which may cause issues if the component is re-rendered with a different entityType
- [ ] Verify the UI renders correctly and matches design expectations
- [ ] Verify team's `parentId` correctly identifies the parent organization for the "disabled by org" badge
- [ ] Test that the badge appears on team features page when org has disabled a feature that team has enabled
- [ ] Verify `TeamRepository.findParentOrganizationByTeamId` returns `{ id: number }` structure as expected by the router
- [ ] **Verify toggle is actually disabled in UI when blocked by higher level (not just showing warning)**
- [ ] **Test that users cannot change feature settings when org has blocked the feature**
- [ ] **Verify policy logic: test permissive vs strict behavior with actual feature configurations**
- [ ] **Verify all callers of `computeEffectiveStateAcrossTeams` pass the required `policy` parameter** (now required, no default)
- [ ] **Review test file rewrite - verify 31 tests provide equivalent coverage to previous implementation**
- [ ] **Verify PBAC procedures in new location (`packages/trpc/server/procedures/pbacProcedures.ts`) are correctly imported**
- [ ] **Verify `featureOptIn.read` and `featureOptIn.update` permissions work correctly** - test that users with these permissions can access/modify feature settings
- [ ] **Verify team features page PBAC implementation** - uses `getServerSession` + direct Prisma membership query + `getResourcePermissions`
- [ ] **Test `canEdit` prop behavior** - verify all toggles (feature toggles AND auto opt-in toggle) are disabled when `canEdit` is false
- [ ] **Note**: `canEdit` defaults to `true` in `FeaturesSettings` - verify this is intentional for user-level features page which doesn't pass the prop

---

Link to Devin run: https://app.devin.ai/sessions/ca2cb0461b494a0bb217a55c5124e830
Requested by: eunjae@cal.com (@eunjae-lee)

devin-ai-integration[bot] committed:
feat: add tri-state UI settings pages and auto opt-in preference

TASK 5: Tri-state UI with ToggleGroup
- Create shared TeamOrgFeaturesSettings component
- Create user features settings page (features-view.tsx)
- Create team features settings page
- Create org features settings page

TASK 6: Add auto opt-in preference
- Add autoOptInExperimentalFeatures field to User and Team models
- Add migration for new fields
- Add TRPC endpoints for auto opt-in preference (using repository pattern)
- Add auto opt-in checkboxes to settings pages

Also adds i18n keys for feature opt-in UI.

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

refactor: rename autoOptInExperimentalFeatures to autoOptInFeatures

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

refactor: apply PBAC procedure pattern to featureOptIn router

- Create createTeamPbacProcedure for team-scoped endpoints
- Create createOrgPbacProcedure for organization-scoped endpoints
- Refactor all team/org endpoints to use the new PBAC procedures
- Significantly reduces boilerplate code for permission checks

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

add features pages

implement hooks

update style

update style

rename slug to featureId

show Features page on the side bar only if OPT_IN_FEATURES.length > 0

revert unncessary renaming

revert some changes

remove some changes

eunjae-lee committed:
fix repository usage

eunjae-lee commented:
@cubic-dev-ai review this PR

devin-ai-integration[bot] committed:
test: add unit tests for PBAC utility procedures

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
feat: add 'disabled by organization' badge for Team Level Features page

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
test: add tests for FeatureOptInService.listFeaturesForTeam and refactor router to use TeamRepository

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
a little clean up

eunjae-lee committed:
feat: add effectiveReason to computeEffectiveStateAcrossTeams for better UI feedback

- Update computeEffectiveStateAcrossTeams to return both enabled state and reason
- Add EffectiveStateReason type with 6 possible values
- Update FeatureOptInService to include effectiveReason in resolved state
- Simplify useUserFeatureOptIn.getBlockedWarning to use effectiveReason directly
- Add feature_no_explicit_enablement_warning translation
- Invalidate feature list when auto-opt-in setting changes in all hooks

eunjae-lee committed:
make it clearer about auto opt in

eunjae-lee committed:
move hooks to apps/web/module

eunjae-lee committed:
fix

sean-brydon reviewed (changes requested):
<img width="2634" height="1086" alt="Image" src="https://github.com/user-attachments/assets/05989dce-df02-4077-8067-d695fd14624b" />

Can't load any page

devin-ai-integration[bot] committed:
Merge main into feature branch, resolve conflict in common.json

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: move types to packages/features/feature-opt-in, fix imports and test mocks

- Move types.ts from apps/web/modules/feature-opt-in/hooks to packages/features/feature-opt-in
- Update FeaturesSettings.tsx to import types from new location
- Update hooks to import types from @calcom/features/feature-opt-in/types
- Fix organization features-view import path (~/settings/organizations -> ~/ee/organizations)
- Fix PermissionCheckService mock in util.test.ts to use class syntax

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
Merge branch 'main' into devin/feature-opt-in-ui-1765816717

eunjae-lee commented on packages/features/feature-opt-in/components/FeaturesSettings.tsx:
```diff
@@ -0,0 +1,136 @@
+import { Tooltip } from "@calcom/ui/components/tooltip";
+
+import { getOptInFeatureConfig } from "../config";
+import type { NormalizedFeature, UseFeatureOptInResult } from "../hooks/types";
```
the import path has been fixed. 

also the type defs live in `packages/features/feature-opt-in/types.ts`

eunjae-lee committed:
Merge branch 'main' into devin/feature-opt-in-ui-1765816717

eunjae-lee committed:
Merge branch 'devin/feature-opt-in-ui-1765816717' of github.com:calcom/cal.com into devin/feature-opt-in-ui-1765816717

devin-ai-integration[bot] committed:
fix: add validation for team ID and remove isPublic from featureOptIn handler

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
Merge branch 'main' into devin/feature-opt-in-ui-1765816717

sean-brydon committed:
Merge branch 'main' into devin/feature-opt-in-ui-1765816717

sean-brydon commented on apps/web/app/(use-page-wrapper)/settings/(settings-layout)/organizations/(org-admin-only)/features/page.tsx:
```diff
@@ -0,0 +1,22 @@
+  );
+
+const Page = async () => {
+  await validateUserHasOrgAdmin();
```
should we use PBAC here? w

devin-ai-integration[bot] committed:
fix: resolve biome lint errors in feature-opt-in files

- Add explicit return types to functions and methods
- Replace ternary operators with if-else statements
- Replace forEach with for...of loops to avoid useIterableCallbackReturn
- Extract helper functions to reduce function complexity
- Add explicit type annotations to variables
- Move exports to end of file to satisfy useExportsLast rule

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: resolve type errors in featureOptIn router

- Replace ctx.organizationId with ctx.user.organization.id
- Add null checks for organizationId in org procedures
- Remove incorrect type annotation from featureOptInRouter

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
refactor: address PR feedback on feature opt-in implementation

- Refactor createMutationCallbacks to useMutationCallbacks hook with useLocale internally
- Restore comment on featureOptIn prop in FeaturesSettings.tsx
- Revert getAllTeamIds helper, use ternary operator instead
- Use ctx.user.organizationId with proper guard clauses

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
fix type error

devin-ai-integration[bot] committed:
feat: disable toggle when feature is blocked by higher level

- Add isBlockedByHigherLevel function to UseFeatureOptInResult interface
- Implement blocking detection for user level (org/team disabled)
- Implement blocking detection for team level (org disabled)
- Organization level never blocked (top level)
- Update FeaturesSettings to disable ToggleGroup when blocked

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
change order of toggle buttons

devin-ai-integration[bot] committed:
feat: add policy field (permissive/strict) to feature opt-in config

- Add OptInFeaturePolicy type with 'permissive' and 'strict' modes
- Update computeEffectiveStateAcrossTeams to handle both policies:
  - Permissive: user opt-in can activate; disables only win if ALL teams disable
  - Strict: user opt-in alone not enough; ANY explicit disable blocks
- Add new EffectiveStateReason values for strict policy
- Update FeatureOptInService to read policy from config
- Add unit tests for all 9 policy scenarios from spec

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
refactor: rewrite test file with scenario tables and make policy required

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
Merge branch 'devin/feature-opt-in-ui-1765816717' of github.com:calcom/cal.com into devin/feature-opt-in-ui-1765816717

eunjae-lee committed:
fix type

eunjae-lee committed:
fix types

eunjae-lee committed:
make policy required

eunjae-lee committed:
disable toggle group if blocked by higher level

devin-ai-integration[bot] committed:
refactor: move PBAC procedures to packages/trpc/server/procedures

- Move util.ts to pbacProcedures.ts in procedures folder for better reusability
- Move util.test.ts to pbacProcedures.test.ts alongside the main file
- Update imports in _router.ts to use new location

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: remove redundant null checks in featureOptIn router

The createOrgPbacProcedure middleware already validates organizationId
and throws if null. Use non-null assertion (!) instead of redundant
runtime checks since the middleware guarantees the value exists.

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
refactor: improve middleware typing to avoid non-null assertions

Remove explicit return type from createOrgPbacProcedure to let TypeScript
infer the extended context type. This allows handlers to use ctx.organizationId
directly (typed as number) instead of ctx.user.organizationId! assertions.

The middleware validates organizationId and adds it to the context, so
downstream handlers can safely access ctx.organizationId as a number.

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
feat(pbac): use featureOptIn permissions instead of generic team/org permissions

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
feat(pbac): add PBAC to team features page and pass canEdit to views

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
Merge branch 'main' into devin/feature-opt-in-ui-1765816717

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: add effectiveReason to IFeatureOptInService and update listFeaturesForTeam interface

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: add feature_any_team_disabled to isUserBlockedByHigherLevel check

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: add strict-policy reasons to warning helper function

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

sean-brydon reviewed (approved):
Overall LGTM - as discussed lets go ahead in a follow up PR implement some sort of Scope config 

export const OPT_IN_FEATURES: OptInFeatureConfig[] = [
  {
     slug: "bookings-v3",
     titleI18nKey: "bookings_v3_title",
     descriptionI18nKey: "bookings_v3_description",
     policy: "permissive",
     scope: [ORG, TEAM, USER]
   },
];

That allows us to scope a feature JUST for a particular usecase as currently all features in the opt in flow show for org,team,users
